File: src/domain/entities/transaction.ts
/**
 * `Transaction` entity — the core record the reversal engine reasons about.
 *
 * This is a domain entity (identity + invariants + behaviour), deliberately
 * decoupled from the persistence shape in `useTransactions`/Supabase. An
 * application-layer mapper is responsible for translating a DB row into this
 * entity; the domain never imports the database client.
 */
import { Money } from "../money";
import { DomainErrorCode, err, ok, type Result } from "../shared/result";
import { PaymentRail, railProfile } from "./payment-rail";

/** Business classification of a money movement. */
export enum TransactionType {
  Deposit = "deposit",
  Withdrawal = "withdrawal",
  Transfer = "transfer",
  /** Currency conversion (two legs, FX spread applied). */
  Fx = "fx",
  Payment = "payment",
}

/** Lifecycle state of a transaction. */
export enum TransactionStatus {
  Initiated = "initiated",
  Confirmed = "confirmed",
  Processing = "processing",
  Completed = "completed",
  Failed = "failed",
  Reversed = "reversed",
}

/** The counterparty channel — affects fraud/dispute base rates. */
export enum CounterpartyType {
  /** Verified merchant with a dispute process. */
  Merchant = "merchant",
  /** Individual / P2P payee. */
  Individual = "individual",
  /** First payment ever sent to this counterparty. */
  FirstParty = "first_party",
  Unknown = "unknown",
}

/** Snapshot of the customer's standing, used to weight risk. */
export interface CustomerRiskSnapshot {
  /** Account age in days. Older accounts are lower risk. */
  readonly accountAgeDays: number;
  /** Lifetime count of successful (non-reversed) transactions. */
  readonly priorSuccessfulTransactions: number;
  /** Count of reversals/chargebacks the customer has previously filed. */
  readonly priorReversals: number;
  /** Whether the customer has completed KYC/identity verification. */
  readonly kycVerified: boolean;
}

export interface TransactionProps {
  readonly id: string;
  readonly type: TransactionType;
  readonly status: TransactionStatus;
  /** Signed amount in minor units; the entity stores it as absolute `Money`. */
  readonly amount: Money;
  /** Settlement currency (for FX, the destination currency). */
  readonly settlementCurrency?: Money["currency"];
  readonly rail: PaymentRail;
  readonly counterpartyType: CounterpartyType;
  /** When the transaction was created. */
  readonly createdAt: Date;
  /** When it settled, if it has. Drives reversal-window math. */
  readonly settledAt?: Date | null;
  /** True when the source and destination currencies differ. */
  readonly isCrossCurrency: boolean;
  readonly customer: CustomerRiskSnapshot;
  /** Free-form metadata carried from upstream systems. */
  readonly metadata?: Readonly<Record<string, unknown>>;
}

export class Transaction {
  private constructor(private readonly props: TransactionProps) {}

  /**
   * Validate invariants and construct a `Transaction`. Returns a `Result`
   * rather than throwing so application code can surface field-level errors.
   */
  static create(props: TransactionProps): Result<Transaction> {
    if (!props.id.trim()) {
      return err(DomainErrorCode.Validation, "Transaction id is required", "id");
    }
    if (props.amount.minor < 0) {
      return err(
        DomainErrorCode.Validation,
        "Transaction amount must be non-negative (use direction, not sign)",
        "amount",
      );
    }
    if (props.amount.isZero) {
      return err(
        DomainErrorCode.Validation,
        "Transaction amount must be greater than zero",
        "amount",
      );
    }
    if (props.settledAt && props.settledAt < props.createdAt) {
      return err(DomainErrorCode.Invariant, "settledAt cannot precede createdAt", "settledAt");
    }
    const profile = railProfile(props.rail);
    if (!profile.supportedCurrencies.includes(props.amount.currency)) {
      return err(
        DomainErrorCode.Validation,
        `Rail ${props.rail} does not support ${props.amount.currency}`,
        "rail",
      );
    }
    return ok(new Transaction(props));
  }

  get id(): string {
    return this.props.id;
  }
  get type(): TransactionType {
    return this.props.type;
  }
  get status(): TransactionStatus {
    return this.props.status;
  }
  get amount(): Money {
    return this.props.amount;
  }
  get rail(): PaymentRail {
    return this.props.rail;
  }
  get counterpartyType(): CounterpartyType {
    return this.props.counterpartyType;
  }
  get customer(): CustomerRiskSnapshot {
    return this.props.customer;
  }
  get isCrossCurrency(): boolean {
    return this.props.isCrossCurrency;
  }
  get createdAt(): Date {
    return this.props.createdAt;
  }
  get settledAt(): Date | null {
    return this.props.settledAt ?? null;
  }
  get metadata(): Readonly<Record<string, unknown>> {
    return this.props.metadata ?? {};
  }

  /** Age of the transaction in hours, measured from `now`. */
  ageHours(now: Date = new Date()): number {
    return (now.getTime() - this.props.createdAt.getTime()) / 3_600_000;
  }

  /** Hours since settlement, or null if not yet settled. */
  hoursSinceSettlement(now: Date = new Date()): number | null {
    const settled = this.settledAt;
    if (!settled) return null;
    return (now.getTime() - settled.getTime()) / 3_600_000;
  }

  /** A transaction already reversed or failed cannot be reversed again. */
  get isTerminalForReversal(): boolean {
    return (
      this.props.status === TransactionStatus.Reversed ||
      this.props.status === TransactionStatus.Failed
    );
  }

  /**
   * Whether the rail's reversal window is still open. Rails without a hard
   * window (e.g. internal book transfers, recalls) are always considered open.
   */
  isWithinReversalWindow(now: Date = new Date()): boolean {
    const profile = railProfile(this.props.rail);
    if (profile.reversalWindowHours === null) return true;
    const sinceSettlement = this.hoursSinceSettlement(now);
    // Not settled yet → window has not started; reversal is still possible.
    if (sinceSettlement === null) return true;
    return sinceSettlement <= profile.reversalWindowHours;
  }

  /**
   * Quick gate: is this transaction eligible to even be analyzed for reversal?
   * The engine still runs richer scoring, but this short-circuits hopeless
   * cases (already reversed, window expired).
   */
  isReversalEligible(now: Date = new Date()): boolean {
    return !this.isTerminalForReversal && this.isWithinReversalWindow(now);
  }
}

File: backend/src/SmartPayEngine.Core/Entities/Transaction.cs
using SmartPayEngine.Core.Enums;
using SmartPayEngine.Core.Exceptions;
using SmartPayEngine.Core.ValueObjects;

namespace SmartPayEngine.Core.Entities;

/// <summary>
/// A money movement on the platform — the central record the reversal engine
/// and routing service reason about.
///
/// The entity is constructed through <see cref="Create"/> so invariants are
/// enforced in one place. Mutable workflow fields (status, completion time) are
/// updated through intention-revealing methods rather than open setters.
/// </summary>
public sealed class Transaction
{
    public Guid Id { get; }

    /// <summary>The principal amount, denominated in <see cref="FromCurrency"/>.</summary>
    public Money Amount { get; }

    /// <summary>Currency funds were debited in.</summary>
    public Currency FromCurrency { get; }

    /// <summary>Currency funds were credited in (differs for FX).</summary>
    public Currency ToCurrency { get; }

    public TransactionStatus Status { get; private set; }

    public DateTimeOffset CreatedAt { get; }

    /// <summary>When settlement completed; <c>null</c> until then.</summary>
    public DateTimeOffset? CompletedAt { get; private set; }

    /// <summary>The provider / rail the transaction was sent over.</summary>
    public PaymentProvider RouteUsed { get; }

    /// <summary>The counterparty channel — influences dispute base rates.</summary>
    public CounterpartyType CounterpartyType { get; }

    /// <summary>
    /// Upstream-supplied risk score in [0, 100] (0 = clean, 100 = certain
    /// fraud). Surfaced as a first-class field because it materially affects
    /// reversal odds; also mirrored into <see cref="Metadata"/> in practice.
    /// </summary>
    public int RiskScore { get; }

    /// <summary>Free-form metadata carried from upstream systems.</summary>
    public IReadOnlyDictionary<string, string> Metadata { get; }

    private Transaction(
        Guid id,
        Money amount,
        Currency fromCurrency,
        Currency toCurrency,
        TransactionStatus status,
        DateTimeOffset createdAt,
        DateTimeOffset? completedAt,
        PaymentProvider routeUsed,
        CounterpartyType counterpartyType,
        int riskScore,
        IReadOnlyDictionary<string, string> metadata)
    {
        Id = id;
        Amount = amount;
        FromCurrency = fromCurrency;
        ToCurrency = toCurrency;
        Status = status;
        CreatedAt = createdAt;
        CompletedAt = completedAt;
        RouteUsed = routeUsed;
        CounterpartyType = counterpartyType;
        RiskScore = riskScore;
        Metadata = metadata;
    }

    /// <summary>
    /// Validate invariants and create a <see cref="Transaction"/>.
    /// </summary>
    /// <exception cref="DomainException">When an invariant is violated.</exception>
    public static Transaction Create(
        Money amount,
        Currency fromCurrency,
        Currency toCurrency,
        PaymentProvider routeUsed,
        TransactionStatus status = TransactionStatus.Completed,
        DateTimeOffset? createdAt = null,
        DateTimeOffset? completedAt = null,
        CounterpartyType counterpartyType = CounterpartyType.Unknown,
        int riskScore = 0,
        IReadOnlyDictionary<string, string>? metadata = null,
        Guid? id = null)
    {
        if (amount.IsZero)
        {
            throw new DomainException("Transaction amount must be greater than zero.", nameof(amount));
        }

        if (amount.Currency != fromCurrency)
        {
            throw new DomainException(
                "Amount currency must match FromCurrency.", nameof(amount));
        }

        if (riskScore is < 0 or > 100)
        {
            throw new DomainException("RiskScore must be between 0 and 100.", nameof(riskScore));
        }

        var created = createdAt ?? DateTimeOffset.UtcNow;
        if (completedAt is { } c && c < created)
        {
            throw new DomainException("CompletedAt cannot precede CreatedAt.", nameof(completedAt));
        }

        return new Transaction(
            id ?? Guid.NewGuid(),
            amount,
            fromCurrency,
            toCurrency,
            status,
            created,
            completedAt,
            routeUsed,
            counterpartyType,
            riskScore,
            metadata ?? new Dictionary<string, string>());
    }

    /// <summary>True when source and destination currencies differ (FX).</summary>
    public bool IsCrossCurrency => FromCurrency != ToCurrency;

    /// <summary>Age of the transaction relative to <paramref name="now"/>.</summary>
    public TimeSpan AgeFrom(DateTimeOffset now) => now - CreatedAt;

    /// <summary>
    /// Reference instant from which a reversal window is measured: completion
    /// time when settled, otherwise creation time.
    /// </summary>
    public DateTimeOffset ReversalAnchor => CompletedAt ?? CreatedAt;

    /// <summary>A transaction already reversed or failed cannot be reversed.</summary>
    public bool IsTerminalForReversal =>
        Status is TransactionStatus.Reversed or TransactionStatus.Failed;

    /// <summary>Mark the transaction settled at <paramref name="at"/>.</summary>
    public void MarkCompleted(DateTimeOffset at)
    {
        Status = TransactionStatus.Completed;
        CompletedAt = at;
    }
}
